### DISCUSSION TITLE
Define requirements for better i18n support

### DISCUSSION THREAD

**Discussion by ramSeraph** [maplibre/maplibre#231] [General]
```
DISCLAIMER: THIS IS A DRAFT DOCUMENT INTENDED TO START A DISCUSSION AND 
MOSTLY REFLECTS MY VIEW OF THE REQUIREMENTS AND IT IS IN NO WAY FINAL OR THE CONSENSUS
```

# Functional Requirements

## Must Have

   - [ BASIC FUNCTIONALITY ] Ability to facilitate basic(legible) rendering of as many scripts/languages as possible, supported list atleast at par with most browsers and UI frameworks.
     -  For a clarification on what constitutes Basic Rendering Support, refer to the `Layout Requirement Priorities` section below.

   - [ BASIC FUNCTIONALITY ] Basic Font fallbacks and ability to use system installed fonts or fonts bundled with native apps and fonts served from remote servers.
     - Basic Font fallback Clarification( / Implementation Guidelines )
        - Font fallbacks should happen at the script-run level.. not at individual char code level.
        - If any of the script-runs can be covered by any of the font fallbacks, the rendering of the text should be made visible as soon as the first suitable fallback is available.

   - [ NON DEGRADATION ] All existing label layout features for the languages already supported by existing implementations of maplibre-gl-\* should be supported.
   
   - [ BACKWARD COMPATIBILITY ] For existing styles using the current glyphs api, all existing layout features and existing font fallback features should be supported.
     - This can be in the main code base or made available as a plugin for the new versions of the code base.

   - [ FUTURE PROOFING ] Serialization formats to be used in font delivery should be loosely coupled with the technologies used for the font rendering implementation on the client side.
     - Well known and well maintained standards or formats which are minimally lossy should be used for the font delivery serialization formats
        -  Well known formats here refer to standard font delivery formats like OTF, TTF, WOFF*
        -  Loss here refers to loss of shaping/styling information
     - Support for specializing font delivery serialization formats, for a particular font rendering implementation, if implemented should be optional, non-default and should be behind a flag

## Good To Have

  - Fonts higher in the fallback priority list, when they become available, should be used to optionally replace labels already rendered/displayed with a font lower in the fallback priority list

  - Ability to support advanced layout features for as many languages/scripts as possible, supported language/script  list being on par with what is supported by existing browsers and UI frameworks, layout features supported being at-least on par with what maplibre-gl-\* can do for the languages it already supports.
    -  For a clarification on what constitutes Advanced Layout Features refer to the `Layout Requirement Priorities` section below.

    - Implementation priorities:
      - ability to with work source data already annotated with markers to help with layout.
        - NOTE: this should probably go into the `Must Have` bucket
      - ability to work with unchanged source text data, which at best has an extra flag for a given symbol-label to indicate locale/language/script.
      - a flag at source/symbol-label level to differentiate between the above two cases.
      - provision of a tool to annotate source data with markers to help with layout

  - Fancy font feature support( like non-essential ligatures )
    - ability to toggle font features at font level
    - ability to toggle font features at symbol-label level
    - documentation on what font features will be switched off for what kind symbol-label( ex: no non-essential ligatures for labels along line )

## Layout Requirement Priorities(/Categories)

   -  [ BASIC ] Simple Point Labels in all languages( no letter spacing, single line )
   -  [ BASIC ] Line Labels in all languages( no letter spacing, single straight line )
   -  [ BASIC ] Support for default writing direction for a script/language
   -  [ BASIC ] Respect default baseline for a script
   -  [ BASIC ] Bidirectional support for mixed text where applicable
   -  [ ADVANCED ] Text transform support ( upper case/lower case etc )
   -  [ ADVANCED ] Multiline text for point labels with simple justification modes
   -  [ ADVANCED ] Multiline text for point labels with balanced justification
   -  [ ADVANCED ] Text along curved path
   -  [ ADVANCED ] Letter spacing support
   -  [ ADVANCED ] Support for non-default writing directions ( Ex: English written vertically )
   -  [ ADVANCED ] For mixed text with different baselines, align text using the baseline info
        
   Some of what is categorized as `ADVANCED` above might be recatogerized as `BASIC` for a given language/script based on ease of implementation

# Non Functional Considerations

  - Impact on initialization time
  - Impact on how quickly label data gets shown when moving around a map( while zooming, panning etc )
  - Impact on how smooth the interactiveness of the map is
  - Impact on amount of data to be transferred
  - Impact on client side memory usage
  - Impact on client side disk space usage
 

# Open Questions
 * If a language/script already supported has a particular feature improved/fixed because of the new implementation, should an optional fallback be provided to match original buggy behaviour?
 * Should the current glyphs api be marked as deprecated and people encouraged to move to the better and newer api? 
 * Should a tool be provided for the above migration where possible?

# CHANGELOG:

Version 0.6 (08/03/2023 17:33 IST):

* Relaxed constraints on font serialization formats by adding the possibility of using minimally lossy formats and clarified definitions of standard font serialization formats and definition of loss

Version 0.5 (07/03/2023 12:20 IST):

* Added script baselines layout requirements
* Added some open questions

Version 0.4 (06/03/2023 14:51 IST):

* Added bidirectional support to the layout requirements

Version 0.3 (06/03/2023 13:13 IST):

* Added writing directions and their categories/priorities to the layout requirements
* Separate Layout requirements/priorities into a completely new section.

Version 0.2 (04/03/2023 21:52 IST):

 * Added note on moving advanced layout features to `Must Have` category for a language based on ease of implementation

**Comment by 1ec5** [comment]
Addressing some of these requirements entails changes outside the style specification proper. For example, I’m not aware of a specification for a fonts API other than whatever would be compatible with [node-fontnik](https://github.com/mapbox/node-fontnik/blob/feeec892515442b4deca53e7ff761721ede5240a/API.md) or [TinySDF](https://github.com/mapbox/tiny-sdf/tree/5bd130204f341acfd974081a791be9397fb10c39#usage) output. Performance considerations will most likely differ between platforms no matter the interface design. But it’s good that you’re thinking about this problem end-to-end.

**Reply by ramSeraph** [reply]
> Addressing some of these requirements entails changes outside the style specification proper.


Oh.. definitely. But for Fonts API ..  TTF, OTF and WOFF is what I was thinking about here.

**Comment by 1ec5** [comment]
> Line Labels in all languages( no letter spacing, single straight line ).

While straight-line rendering would be better than nothing for Indic scripts, it would represent a significant regression for other scripts that don’t require complex text layout. I’m not confident that the user base would accept a tradeoff between the two. Either we would need to find some solution for curved text in all scripts, or we’d need to implement a parallel code path for the scripts that we can’t curve for now.

**Reply by ramSeraph** [reply]
I think, in that case the definition of "Basic Rendering Support" should be a priority list where we support as much as we can for a given language and what can't be supported goes into the "Good to Have" list. BTW.. I think "letter spacing" is going to be a bigger problem for Indic scripts than "curved text on a line"

**Reply by 1ec5** [reply]
Sure, though I think there would be concern about any solution that results in _Latin_ text becoming uncurved.

**Reply by ramSeraph** [reply]
I think that part is covered in the `NON DEGRADATION` part of the requirements.

**Comment by wipfli** [comment]
Thanks a lot for writing this @ramSeraph. To be as good as modern browsers is maybe a bit too ambitious. For example, we might not need emoji support, or also we might not need line breaking. What do you think?

**Reply by ramSeraph** [reply]
Modern browsers have more problems to solve than we do.. like text editing and cursor movement. We might actually be able to do better, since we will end up using the same tools they are using to solve this problem but with lesser constraints. About line-breaking.. I think it's already something that we are doing for the languages supported. @1ec5 had something to say about emoji support before and unicode supplementary planes and the need to support them( I didn't understand it :), but I think it might be needed ). 

EDIT: I should have waited for @1ec5 to comment. :)

**Reply by 1ec5** [reply]
Support for complex text layout would most likely solve one of the blockers for monochrome emoji (the codepoint=glyph assumption) and possibly another (8 bits per character): https://github.com/mapbox/mapbox-gl-native/issues/15666#issuecomment-532969076. The third blocker is specific to Mapbox’s font selection. Color emoji is another challenge altogether, but I agree that it’s less important than proper internationalization.

**Comment by ramSeraph** [comment]
A very interesting and relevant study on remote font delivery  - https://www.w3.org/TR/2020/NOTE-PFE-evaluation-20201015/

**Reply by ramSeraph** [reply]
Actually.. Let's not discuss solutions here yet. Let's aim for consensus on requirements. Do read, but let's not discuss that here.

**Comment by wipfli** [comment]
I would like to add `[ BASIC ] Bidirectional support` to the layout requirements.

**Reply by ramSeraph** [reply]
Done :)
